framework/aster-frame/src/vm/page_table/cursor.rs

```diff
@@ -0,0 +1,529 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::{any::TypeId, mem::size_of, ops::Range};
+
+use super::{
+    Child, KernelMode, MapInfo, MapOp, MapProperty, PageTable, PageTableConstsTrait,
+    PageTableEntryTrait, PageTableError, PageTableFrame, PageTableMode, PtfRef,
+};
+use crate::{
+    sync::{ArcSpinLockGuard, SpinLock},
+    vm::{paddr_to_vaddr, Paddr, Vaddr, VmFrame},
+};
+
+/// The cursor for forward traversal over the page table.
+///
+/// Each method may move the cursor forward, doing mapping unmaping, or
+/// querying this slot.
+///
+/// Doing mapping is somewhat like a depth-first search on a tree, except
+/// that we modify the tree while traversing it. We use a stack to simulate
+/// the recursion.
+///
+/// During traversal, if we want to have modifications on the current node,
+/// we will try to aquire the lock of the tree from root to the current
+/// node.
```

Comment:
Well, what is the rationale for this? Is this an existing locking protocol or something similar to the Linux kernel implementation? I cannot understand it, so is there some other justification?

Reply:
It seems that the documentation is malformed. What is implemented is: any read or write accesses to nodes require exclusive access on the entire path from the root to the node.

It is a doc of a previous implementation, I'll update it.

As for rationale for the current implementation, I just think it is true,  sadly no other supporting materials.

Reply:
```rust
/// Any read or write accesses to nodes require exclusive access on the
/// entire path from the root to the node. But cursor can be created without
/// holding the lock, and can release the lock after yeilding the current
/// slot while querying over the page table with a range. Simultaneous
/// reading or writing to the same range in the page table will not produce
/// consistent results, only validity is guaranteed.
```

Here's the updated doc. Since the cursor is an inner implementer of map/unmap/query operations and cursors are created in controlled contexts, I want to ensure consistency in higher abstraction levels (such as holding extra range locks in `map()`) since currently the implementation of cursor is sound.

Reply:
> What is implemented is: any read or write accesses to nodes require exclusive access on the entire path from the root to the node.

What I mean is that I don't think it's right. It's not right even if we forbid "reading or writing to the same range in the page table".

For example, suppose the last level page table looks like this:
```
---------*-----------------------------------------------
         ^                                  ^
         |                                  |
     PTE x (mapped)                     PTE y (empty)
```

Now, the user program unmaps `PTE x` and maps `PTE y` simultaneously.
 1. First, we create a cursor for `PTE x` and another cursor `PTE y`.
 2. The second cursor unmaps `PTE y` first. Since the last level page table becomes empty, it gets released.
 3. However, a reference to the last level page table can still exist in the first cursor (in `stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],`), which will cause problems. 

Note that, `acquire_locks` and `release_locks` have nothing to do with this problem. They acquire the locks, but do not ensure all the locked page tables are still alive.

What I still want to see is to implement a proper lock protocol. For example, I can suggest the followings:
 - `PageTableCursor::new` should accept an address range, which represents that the cursor will not (and cannot) touch page table entries that represent virtual addresses outside that range.
   - Then, `PageTableCursor::new` finds an intermediate page table (not necessarily the last-level or the top-level) which represents an address range that contains the whole specified address range. It requires all locks from the root page table to the intermediate page table, but then unlocks all locks excluding the one for the intermediate page table.
 - `PageTableCursor` then maintains the lock guards from one for the intermediate page table to the leaf. It shouldn't have `stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],` anymore. Instead, it should only have `lock_guard: [Option<ArcSpinLockGuard<PageTableFrame<E, C>>>; C::NR_LEVELS],`, which starts from the one for the intermediate page table to the leaf page table.

For example, if we're going to map the address range shown below:
```
Top-level page tables             A
                                 /
                                B
                               / \
Last-level page tables        C   D
Last-level PTEs           ---**   **---
                             ^     ^
                             |     |
                             +--+--+
                                |
                Address range that we're going to map
```
 - When calling `PageTableCursor`, it will:
   - `lock(A)`, `lock(B)`, `unlock(A)`
   - `lock_guard = [ locked(B) ]`
 - When calling `PageTableCursor::map`, it will:
   - `lock(C)`, `lock_guard = [ locked(B), locked(C) ]`
   - Map some pages in `C`
   - `unlock(C)`, `lock_guard = [ locked(B) ]`
   - `lock(D)`, `lock_guard = [ locked(B), locked(D) ]`
   - Map some pages in `D`
   - `unlock(D)`, `lock_guard = [ locked(D) ]`

How do you like this idea? AFAIK, it should be deadlock-free and race-condition-free, even if we make `PageTableCursor` public and allow two cursors to operate on the same address range.

> Since the cursor is an inner implementer of map/unmap/query operations and cursors are created in controlled contexts, I want to ensure consistency in higher abstraction levels (such as holding extra range locks in `map()`) since currently the implementation of cursor is sound.

I don't think we hold any locks in the current implementation of `map()`, do we? Even worse, if we actually hold a lock in `map()`, additional locks in the cursor are pointless, right?

Reply:
> How do you like this idea?

Indeed, I just realize one problem. With this approach, we cannot easily release empty page tables up to the root page table.
 - In the above example, even if `B` becomes empty, we cannot directly release it. It is because while holding the lock of `B`, we cannot relock `A`. Otherwise, it may cause deadlocks.

This can be fixed by adding another step. If the intermediate page table `B` becomes empty, we will first release its lock. Then, we try to acquire the locks from the root page table to the intermediate page table again (i.e., `lock(A)`, `lock(B)`) and recheck whether `B` is still empty. By doing so, we can easily release empty pages along the path.

I think it's acceptable and I cannot figure out a better way to resolve it.
 - An alternative is to hold all locks from the root page table (i.e., `A`), not the intermediate page table (i.e., `B`), to the leaf page table in the cursor. But it's pointless because it means that there is a big lock in the root page table and all other locks are pointless.
 - The root cause is that we cannot hold page tables (in `stack`) without holding their locks (in `stack_guard`). Otherwise, we'll have race conditions. So we should hold only page table lock guards `stack_guard` and remove `stack`. However, it means that we lost track to the unlocked pages without trying to relock them from the root page table again.

Reply:
Marvelous. Indeed the current implementation could render another cursor invalid if a cursor unmaps a PTE, which is not as valid as I thought it could be.

Give me some time trying out the protocol you proposed. Thanks.

Reply:
Well, here's my plan fixing this:
 - The cursor only hold lock guards, as you suggested;
 - When creating a cursor, a range is needed and the cursor will ensure exclusive ownership over that range. This is done by holding a lock of a page table containing that range, as you suggested;
 - However, when the cursor is created, it only drop the lock on the root node if applicable, allowing us for at most 512 ready cursors at the same time;
 - 2nd level page tables won't be freed by a cursor unless explicity recycled by another method. That introduces minimal wastage, but the wastage is desired for the kernel part(upper 256).

Reply:
> * However, when the cursor is created, it only drop the lock on the root node if applicable, allowing us for at most 512 ready cursors at the same time;
> 
> * 2nd level page tables won't be freed by a cursor unless explicity recycled by another method. That introduces minimal wastage, but the wastage is desired for the kernel part(upper 256).

That does sound right.

The only potential concern is that it prevents concurrent access to the page table if both virtual addresses fall into the same entry of the second-level page table. So I guess you prefer this approach mainly for code simplicity?

Assume that the answer to the above question is yes. I'm not sure how well it does the job, since re-walking the page table to free the intermediate page table doesn't look very complicated either (and does not bring much runtime overhead since it only occurs if the intermediate page table becomes empty).

Reply:
The problem with the rewalk is that you need to hold the reference to the root node just like what you did with the stack.

Maybe we can prove it correct even for the copy-on-write page table. As long as we mark cursors as RW cursors and Read only cursors.

Reply:
Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

I haven't found a solution to it, since the pending to be created cursor will still be locked by another and the creator cannot tell if it is pending to be recycled in previous modifications to it.

Reply:
> The problem with the rewalk is that you need to hold the reference to the root node just like what you did with the stack.

Well, the page table root is special. Because the root should never be deallocated (unless the whole page table is deallocated, of course) or replaced, should it?

If you do not want to keep a direct reference to the page table root, you can also reference the page table itself, i.e. `&'a PageTable`. This will make the cursor look even more correct (as usual, a cursor must have a lifetime parameter to prevent unexpected outlives).

By the way, I don't understand why `PageTableCursor` has an unused lifetime parameter `'a` in the current implementation. Why doesn't the compiler give a warning to make the CI fail?

```rust
pub(super) struct PageTableCursor<
    'a,
    M: PageTableMode,
    E: PageTableEntryTrait,
    C: PageTableConstsTrait,
> where
    [(); C::NR_ENTRIES_PER_FRAME]:,
    [(); C::NR_LEVELS]:,
{
    stack: [Option<PtfRef<E, C>>; C::NR_LEVELS],
    lock_guard: [Option<ArcSpinLockGuard<PageTableFrame<E, C>>>; C::NR_LEVELS],
    level: usize,
    va: Vaddr,
}
```

> Maybe we can prove it correct even for the copy-on-write page table. As long as we mark cursors as RW cursors and Read only cursors.

COW page tables look much more complex. Honestly speaking, I haven't given it much thought.

Reply:
> Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

Ok, I know how. You need to release the lock of B before rewalking, and check the child counter again when reached B.

Reply:
> Well I pondered upon it again. I think you are right. But it is a little bit trickier to avoid deadlocks: If a cursor has the lock of B and wants to recycle B, and another one tries to create a cursor mapping in the range of B, there would be a deadlock.

I mean that we should release the lock, re-walk the page table from the root, and re-check whether `B` is still empty, as mentioned in https://github.com/asterinas/asterinas/pull/713#discussion_r1582009463:

> This can be fixed by adding another step. If the intermediate page table `B` becomes empty, we will first release its lock. Then, we try to acquire the locks from the root page table to the intermediate page table again (i.e., `lock(A)`, `lock(B)`) and recheck whether `B` is still empty. By doing so, we can easily release empty pages along the path.

I think doing that is okay because it happens _only_ when we drop the cursor and _only_ if the intermediate page table becomes empty in the first place. The overhead is acceptable. Meanwhile, it gives us opportunities to release not only `B` but also all `B`'s parents.